extensions/common/context.cc

```diff
@@ -295,13 +277,75 @@ flatbuffers::DetachedBuffer extractLocalNodeFlatBuffer() {
   return fbb.Release();
 }
 
+PeerNodeInfo::PeerNodeInfo(const std::string_view peer_metadata_id_key,
+                           const std::string_view peer_metadata_key) {
+  found_ = getValue({peer_metadata_id_key}, &peer_id_);
+  if (found_) {
+    getValue({peer_metadata_key}, &peer_node_);
+    return;
+  }
+  if (peer_metadata_id_key == kDownstreamMetadataIdKey) {
+    // Downstream peer's metadata will never be in localhost endpoint. Skip
+    // looking for it.
+    return;
+  }
+
+  // Construct a fallback peer node metadata based on endpoint labels if it is
+  // not in filter state.
+  std::string endpoint_labels;
+  fallback_peer_node_ = extractEmptyNodeFlatBuffer();
+  if (!getValue(
+          {"upstream_host_metadata", "filter_metadata", "istio", "workload"},
+          &endpoint_labels)) {
+    return;
+  }
+  std::vector<std::string_view> parts = absl::StrSplit(endpoint_labels, ';');
+  // workload label should semicolon separated four parts string:
+  // workload_name;namespace;canonical_service;canonical_revision.
+  if (parts.size() < 4) {
+    return;
+  }
+
+  flatbuffers::FlatBufferBuilder fbb;
+  flatbuffers::Offset<flatbuffers::String> workload_name, namespace_;
+  std::vector<flatbuffers::Offset<KeyVal>> labels;
+  workload_name = fbb.CreateString(parts[0]);
+  namespace_ = fbb.CreateString(parts[1]);
+  if (!parts[2].empty()) {
+    labels.push_back(CreateKeyVal(fbb,
+                                  fbb.CreateString(kCanonicalServiceLabelName),
+                                  fbb.CreateString(parts[2])));
+  }
+  if (!parts[3].empty()) {
+    labels.push_back(
+        CreateKeyVal(fbb, fbb.CreateString(kCanonicalServiceRevisionLabelName),
+                     fbb.CreateString(parts[3])));
+  }
+  auto labels_offset = fbb.CreateVectorOfSortedTables(&labels);
+
+  FlatNodeBuilder node(fbb);
+  node.add_workload_name(workload_name);
+  node.add_namespace_(namespace_);
+  node.add_labels(labels_offset);
+  auto data = node.Finish();
+  fbb.Finish(data);
+  fallback_peer_node_ = fbb.Release();
```

Comment:
should found_ be updated here? we don't want to wait for metadata exchange in this case?

Reply:
found_ here means finding it from filter metadata (written by metadata exchange filter). I think we'd still want to wait for metadata exchange to happen, as the fallback peer node only has limited information (workload name and namespace, canonical labels) due to performance concern.

Reply:
sounds good